File: packages/vmcp-dsl/src/patterns/stateful.ts
/**
 * Stateful pattern builders
 *
 * These patterns require external state stores and are not yet implemented
 * in the runtime. The IR is defined so compositions can be authored and
 * validated, with helpful errors when execution is attempted.
 */

import type {
  PatternSpec,
  StepOperation,
  RetrySpec,
  TimeoutSpec,
  CacheSpec,
  IdempotentSpec,
  CircuitBreakerSpec,
  DeadLetterSpec,
  SagaSpec,
  SagaStep,
  ClaimCheckSpec,
  BackoffStrategy,
  FieldPredicate,
  DataBinding,
  OnDuplicate,
} from '../types.js';

// =============================================================================
// Duration helpers
// =============================================================================

export class Duration {
  private constructor(private ms: number) {}

  static millis(n: number): Duration {
    return new Duration(n);
  }
  static seconds(n: number): Duration {
    return new Duration(n * 1000);
  }
  static minutes(n: number): Duration {
    return new Duration(n * 60 * 1000);
  }
  static hours(n: number): Duration {
    return new Duration(n * 60 * 60 * 1000);
  }

  toMillis(): number {
    return this.ms;
  }
  toSeconds(): number {
    return Math.floor(this.ms / 1000);
  }
}

export const millis = Duration.millis;
export const seconds = Duration.seconds;
export const minutes = Duration.minutes;
export const hours = Duration.hours;

// =============================================================================
// Backoff strategies
// =============================================================================

export function fixedBackoff(delay: Duration): BackoffStrategy {
  return { fixed: { delayMs: delay.toMillis() } };
}

export function exponentialBackoff(
  initialDelay: Duration,
  maxDelay: Duration,
  multiplier: number = 2
): BackoffStrategy {
  return {
    exponential: {
      initialDelayMs: initialDelay.toMillis(),
      maxDelayMs: maxDelay.toMillis(),
      multiplier,
    },
  };
}

export function linearBackoff(
  initialDelay: Duration,
  increment: Duration,
  maxDelay: Duration
): BackoffStrategy {
  return {
    linear: {
      initialDelayMs: initialDelay.toMillis(),
      incrementMs: increment.toMillis(),
      maxDelayMs: maxDelay.toMillis(),
    },
  };
}

// =============================================================================
// Retry pattern
// =============================================================================

export interface RetryConfig {
  /** The tool or pattern to retry */
  inner: StepOperation;
  /** Maximum number of attempts (including initial) */
  maxAttempts: number;
  /** Backoff strategy between attempts */
  backoff: BackoffStrategy;
  /** Condition to retry (if absent, retry all errors) */
  retryIf?: FieldPredicate;
  /** Jitter factor (0-1) to add randomness */
  jitter?: number;
  /** Per-attempt timeout */
  attemptTimeout?: Duration;
}

/**
 * Create a retry wrapper around a tool or pattern.
 *
 * @example
 * retry({
 *   inner: { tool: { name: 'unreliable_api' } },
 *   maxAttempts: 3,
 *   backoff: exponentialBackoff(millis(100), seconds(5), 2),
 *   jitter: 0.1,
 * })
 */
export function retry(config: RetryConfig): PatternSpec {
  const spec: RetrySpec = {
    inner: config.inner,
    maxAttempts: config.maxAttempts,
    backoff: config.backoff,
    retryIf: config.retryIf,
    jitter: config.jitter,
    attemptTimeoutMs: config.attemptTimeout?.toMillis(),
  };
  return { retry: spec };
}

// =============================================================================
// Timeout pattern
// =============================================================================

export interface TimeoutConfig {
  /** The tool or pattern to wrap */
  inner: StepOperation;
  /** Maximum duration before timeout */
  duration: Duration;
  /** Fallback on timeout (optional) */
  fallback?: StepOperation;
  /** Custom error message */
  message?: string;
}

/**
 * Create a timeout wrapper around a tool or pattern.
 *
 * @example
 * timeout({
 *   inner: { tool: { name: 'slow_tool' } },
 *   duration: seconds(30),
 *   fallback: { tool: { name: 'cached_fallback' } },
 * })
 */
export function timeout(config: TimeoutConfig): PatternSpec {
  const spec: TimeoutSpec = {
    inner: config.inner,
    durationMs: config.duration.toMillis(),
    fallback: config.fallback,
    message: config.message,
  };
  return { timeout: spec };
}

// =============================================================================
// Cache pattern
// =============================================================================

export interface CacheConfig {
  /** JSONPath expressions to derive cache key */
  keyPaths: string[];
  /** The tool or pattern to cache */
  inner: StepOperation;
  /** Store reference name (configured in gateway) */
  store: string;
  /** Time-to-live for cached entries */
  ttl: Duration;
  /** Stale-while-revalidate window */
  staleWhileRevalidate?: Duration;
  /** Condition to cache (if absent, always cache) */
  cacheIf?: FieldPredicate;
}

/**
 * Create a caching wrapper around a tool or pattern.
 *
 * @example
 * cache({
 *   keyPaths: ['$.query', '$.filters.category'],
 *   inner: { tool: { name: 'web_search' } },
 *   store: 'result_cache',
 *   ttl: minutes(15),
 * })
 */
export function cache(config: CacheConfig): PatternSpec {
  const spec: CacheSpec = {
    keyPaths: config.keyPaths,
    inner: config.inner,
    store: config.store,
    ttlSeconds: config.ttl.toSeconds(),
    staleWhileRevalidateSeconds: config.staleWhileRevalidate?.toSeconds(),
    cacheIf: config.cacheIf,
  };
  return { cache: spec };
}

// =============================================================================
// Idempotent pattern
// =============================================================================

export interface IdempotentConfig {
  /** JSONPath expressions to derive idempotency key */
  keyPaths: string[];
  /** The tool or pattern to wrap */
  inner: StepOperation;
  /** Store reference name (configured in gateway) */
  store: string;
  /** Time-to-live for seen-set entry */
  ttl?: Duration;
  /** Behavior on duplicate (default: 'cached') */
  onDuplicate?: OnDuplicate;
}

/**
 * Create an idempotent wrapper to prevent duplicate processing.
 *
 * @example
 * idempotent({
 *   keyPaths: ['$.documentId', '$.operation'],
 *   inner: { tool: { name: 'expensive_analysis' } },
 *   store: 'idempotency_store',
 *   ttl: hours(1),
 * })
 */
export function idempotent(config: IdempotentConfig): PatternSpec {
  const spec: IdempotentSpec = {
    keyPaths: config.keyPaths,
    inner: config.inner,
    store: config.store,
    ttlSeconds: config.ttl?.toSeconds(),
    onDuplicate: config.onDuplicate ?? 'cached',
  };
  return { idempotent: spec };
}

// =============================================================================
// Circuit Breaker pattern
// =============================================================================

export interface CircuitBreakerConfig {
  /** Unique name for this circuit (for state isolation) */
  name: string;
  /** The tool or pattern to protect */
  inner: StepOperation;
  /** Store for circuit state */
  store: string;
  /** Number of failures to trip the circuit */
  failureThreshold: number;
  /** Window for counting failures */
  failureWindow: Duration;
  /** Time to wait before half-open */
  resetTimeout: Duration;
  /** Successes needed in half-open to close (default: 1) */
  successThreshold?: number;
  /** Fallback when circuit is open */
  fallback?: StepOperation;
  /** Custom failure condition */
  failureIf?: FieldPredicate;
}

/**
 * Create a circuit breaker around a tool or pattern.
 *
 * @example
 * circuitBreaker({
 *   name: 'payment_api',
 *   inner: { tool: { name: 'payment_service' } },
 *   store: 'circuit_state',
 *   failureThreshold: 5,
 *   failureWindow: minutes(1),
 *   resetTimeout: seconds(30),
 *   fallback: { tool: { name: 'queue_for_later' } },
 * })
 */
export function circuitBreaker(config: CircuitBreakerConfig): PatternSpec {
  const spec: CircuitBreakerSpec = {
    name: config.name,
    inner: config.inner,
    store: config.store,
    failureThreshold: config.failureThreshold,
    failureWindowSeconds: config.failureWindow.toSeconds(),
    resetTimeoutSeconds: config.resetTimeout.toSeconds(),
    successThreshold: config.successThreshold,
    fallback: config.fallback,
    failureIf: config.failureIf,
  };
  return { circuitBreaker: spec };
}

// =============================================================================
// Dead Letter pattern
// =============================================================================

export interface DeadLetterConfig {
  /** The tool or pattern to wrap */
  inner: StepOperation;
  /** Tool to invoke on failure */
  deadLetterTool: string;
  /** Max attempts before dead-lettering (default: 1) */
  maxAttempts?: number;
  /** Backoff between attempts */
  backoff?: BackoffStrategy;
  /** Whether to rethrow after dead-lettering */
  rethrow?: boolean;
}

/**
 * Create a dead letter wrapper to capture failures.
 *
 * @example
 * deadLetter({
 *   inner: { tool: { name: 'important_process' } },
 *   deadLetterTool: 'failed_jobs_queue',
 *   maxAttempts: 3,
 *   backoff: exponentialBackoff(millis(100), seconds(5)),
 * })
 */
export function deadLetter(config: DeadLetterConfig): PatternSpec {
  const spec: DeadLetterSpec = {
    inner: config.inner,
    deadLetterTool: config.deadLetterTool,
    maxAttempts: config.maxAttempts,
    backoff: config.backoff,
    rethrow: config.rethrow,
  };
  return { deadLetter: spec };
}

// =============================================================================
// Saga pattern
// =============================================================================

export interface SagaStepConfig {
  /** Step name for tracing/logging */
  name: string;
  /** The action to perform */
  action: StepOperation;
  /** Compensating action if later steps fail */
  compensate?: StepOperation;
  /** Input binding for this step */
  input?: DataBinding;
}

export interface SagaConfig {
  /** Ordered list of saga steps */
  steps: SagaStepConfig[];
  /** Store for saga state (for recovery) */
  store?: string;
  /** JSONPath to derive saga instance ID */
  sagaIdPath?: string;
  /** Timeout for entire saga */
  timeout?: Duration;
  /** Output binding */
  output?: DataBinding;
}

/**
 * Create a saga (distributed transaction with compensation).
 *
 * @example
 * saga({
 *   sagaIdPath: '$.orderId',
 *   steps: [
 *     {
 *       name: 'reserve_inventory',
 *       action: { tool: { name: 'reserve_inventory' } },
 *       compensate: { tool: { name: 'release_inventory' } },
 *     },
 *     {
 *       name: 'charge_payment',
 *       action: { tool: { name: 'charge_payment' } },
 *       compensate: { tool: { name: 'refund_payment' } },
 *     },
 *   ],
 *   store: 'saga_state',
 *   timeout: minutes(5),
 * })
 */
export function saga(config: SagaConfig): PatternSpec {
  const steps: SagaStep[] = config.steps.map((step, idx) => ({
    id: `saga_step_${idx}`,
    name: step.name,
    action: step.action,
    compensate: step.compensate,
    input: step.input ?? { input: { path: '$' } },
  }));

  const spec: SagaSpec = {
    steps,
    store: config.store,
    sagaIdPath: config.sagaIdPath,
    timeoutMs: config.timeout?.toMillis(),
    output: config.output,
  };
  return { saga: spec };
}

// =============================================================================
// Claim Check pattern
// =============================================================================

export interface ClaimCheckConfig {
  /** Tool to store payload and return reference */
  storeTool: string;
  /** Tool to retrieve payload from reference */
  retrieveTool: string;
  /** Inner pattern operating on reference */
  inner: StepOperation;
  /** Whether to retrieve original at end */
  retrieveAtEnd?: boolean;
}

/**
 * Create a claim check pattern for large payload handling.
 *
 * @example
 * claimCheck({
 *   storeTool: 'blob_store_put',
 *   retrieveTool: 'blob_store_get',
 *   inner: { pattern: pipeline([...]) },
 *   retrieveAtEnd: true,
 * })
 */
export function claimCheck(config: ClaimCheckConfig): PatternSpec {
  const spec: ClaimCheckSpec = {
    storeTool: config.storeTool,
    retrieveTool: config.retrieveTool,
    inner: config.inner,
    retrieveAtEnd: config.retrieveAtEnd,
  };
  return { claimCheck: spec };
}

File: crates/agentgateway/src/mcp/registry/patterns/stateful.rs
// Stateful pattern type definitions for tool compositions
//
// These patterns require external state stores and are not yet implemented
// in the runtime. The IR types are defined so compositions can be parsed
// and validated, with helpful errors when execution is attempted.

use super::{DataBinding, FieldPredicate, StepOperation};
use serde::{Deserialize, Serialize};

// =============================================================================
// Retry Pattern
// =============================================================================

/// RetrySpec - retry with configurable backoff on failure
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RetrySpec {
    /// The operation to retry
    pub inner: Box<StepOperation>,

    /// Maximum attempts (including initial)
    pub max_attempts: u32,

    /// Backoff strategy
    pub backoff: BackoffStrategy,

    /// Condition to retry (if absent, retry all errors)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub retry_if: Option<FieldPredicate>,

    /// Jitter factor (0.0 - 1.0)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub jitter: Option<f32>,

    /// Per-attempt timeout in milliseconds
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub attempt_timeout_ms: Option<u32>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum BackoffStrategy {
    Fixed(FixedBackoff),
    Exponential(ExponentialBackoff),
    Linear(LinearBackoff),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FixedBackoff {
    pub delay_ms: u32,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExponentialBackoff {
    pub initial_delay_ms: u32,
    pub max_delay_ms: u32,
    #[serde(default = "default_multiplier")]
    pub multiplier: f32,
}

fn default_multiplier() -> f32 {
    2.0
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LinearBackoff {
    pub initial_delay_ms: u32,
    pub increment_ms: u32,
    pub max_delay_ms: u32,
}

// =============================================================================
// Timeout Pattern
// =============================================================================

/// TimeoutSpec - enforce maximum execution duration
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TimeoutSpec {
    /// The operation to wrap
    pub inner: Box<StepOperation>,

    /// Timeout duration in milliseconds
    pub duration_ms: u32,

    /// Fallback on timeout (optional)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fallback: Option<Box<StepOperation>>,

    /// Custom error message
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

// =============================================================================
// Cache Pattern
// =============================================================================

/// CacheSpec - read-through caching with TTL
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CacheSpec {
    /// JSONPath expressions to derive cache key
    pub key_paths: Vec<String>,

    /// The operation to cache
    pub inner: Box<StepOperation>,

    /// Store reference name (configured in gateway)
    pub store: String,

    /// TTL in seconds
    pub ttl_seconds: u32,

    /// Stale-while-revalidate window in seconds
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub stale_while_revalidate_seconds: Option<u32>,

    /// Condition to cache result (if absent, always cache)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cache_if: Option<FieldPredicate>,
}

// =============================================================================
// Idempotent Pattern
// =============================================================================

/// IdempotentSpec - prevent duplicate processing
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IdempotentSpec {
    /// JSONPath expressions to derive idempotency key
    pub key_paths: Vec<String>,

    /// The operation to wrap
    pub inner: Box<StepOperation>,

    /// Store reference name (configured in gateway)
    pub store: String,

    /// TTL in seconds (None = no expiry)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ttl_seconds: Option<u32>,

    /// Behavior on duplicate
    #[serde(default)]
    pub on_duplicate: OnDuplicate,
}

#[derive(Debug, Clone, Copy, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum OnDuplicate {
    #[default]
    Cached,
    Skip,
    Error,
}

// =============================================================================
// Circuit Breaker Pattern
// =============================================================================

/// CircuitBreakerSpec - fail fast with automatic recovery
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CircuitBreakerSpec {
    /// Unique name for this circuit (for state isolation)
    pub name: String,

    /// The protected operation
    pub inner: Box<StepOperation>,

    /// Store for circuit state
    pub store: String,

    /// Number of failures to trip the circuit
    pub failure_threshold: u32,

    /// Window for counting failures (seconds)
    pub failure_window_seconds: u32,

    /// Time to wait before half-open (seconds)
    pub reset_timeout_seconds: u32,

    /// Successes needed in half-open to close (default: 1)
    #[serde(default = "default_success_threshold")]
    pub success_threshold: u32,

    /// Fallback when circuit is open (optional)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fallback: Option<Box<StepOperation>>,

    /// Custom failure condition (if absent, any error)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub failure_if: Option<FieldPredicate>,
}

fn default_success_threshold() -> u32 {
    1
}

// =============================================================================
// Dead Letter Pattern
// =============================================================================

/// DeadLetterSpec - capture failures for later processing
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeadLetterSpec {
    /// The operation to wrap
    pub inner: Box<StepOperation>,

    /// Tool to invoke on failure
    pub dead_letter_tool: String,

    /// Max attempts before dead-lettering (default: 1)
    #[serde(default = "default_max_attempts")]
    pub max_attempts: u32,

    /// Backoff between attempts
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub backoff: Option<BackoffStrategy>,

    /// Whether to rethrow after dead-lettering
    #[serde(default)]
    pub rethrow: bool,
}

fn default_max_attempts() -> u32 {
    1
}

// =============================================================================
// Saga Pattern
// =============================================================================

/// SagaSpec - distributed transaction with compensation
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SagaSpec {
    /// Ordered list of saga steps
    pub steps: Vec<SagaStep>,

    /// Store for saga state (for recovery)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub store: Option<String>,

    /// JSONPath to derive saga instance ID
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub saga_id_path: Option<String>,

    /// Timeout for entire saga in milliseconds
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout_ms: Option<u32>,

    /// Output binding
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output: Option<DataBinding>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SagaStep {
    /// Step identifier
    pub id: String,

    /// Human-readable name
    pub name: String,

    /// The action to perform
    pub action: StepOperation,

    /// Compensating action (optional)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub compensate: Option<StepOperation>,

    /// Input binding for this step
    pub input: DataBinding,
}

// =============================================================================
// Claim Check Pattern
// =============================================================================

/// ClaimCheckSpec - externalize large payloads
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClaimCheckSpec {
    /// Tool to store payload and return reference
    pub store_tool: String,

    /// Tool to retrieve payload from reference
    pub retrieve_tool: String,

    /// Inner operation operating on reference
    pub inner: Box<StepOperation>,

    /// Whether to retrieve original at end
    #[serde(default)]
    pub retrieve_at_end: bool,
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_retry_spec() {
        let json = r#"{
            "inner": { "tool": { "name": "flaky_api" } },
            "maxAttempts": 3,
            "backoff": {
                "exponential": {
                    "initialDelayMs": 100,
                    "maxDelayMs": 5000,
                    "multiplier": 2.0
                }
            },
            "jitter": 0.1
        }"#;

        let spec: RetrySpec = serde_json::from_str(json).unwrap();
        assert_eq!(spec.max_attempts, 3);
        assert!(matches!(spec.backoff, BackoffStrategy::Exponential(_)));
        assert_eq!(spec.jitter, Some(0.1));
    }

    #[test]
    fn test_parse_timeout_spec() {
        let json = r#"{
            "inner": { "tool": { "name": "slow_tool" } },
            "durationMs": 30000,
            "message": "Operation timed out"
        }"#;

        let spec: TimeoutSpec = serde_json::from_str(json).unwrap();
        assert_eq!(spec.duration_ms, 30000);
        assert_eq!(spec.message, Some("Operation timed out".to_string()));
    }

    #[test]
    fn test_parse_cache_spec() {
        let json = r#"{
            "keyPaths": ["$.query", "$.filters"],
            "inner": { "tool": { "name": "search" } },
            "store": "result_cache",
            "ttlSeconds": 900
        }"#;

        let spec: CacheSpec = serde_json::from_str(json).unwrap();
        assert_eq!(spec.key_paths.len(), 2);
        assert_eq!(spec.store, "result_cache");
        assert_eq!(spec.ttl_seconds, 900);
    }

    #[test]
    fn test_parse_circuit_breaker_spec() {
        let json = r#"{
            "name": "payment_api",
            "inner": { "tool": { "name": "payment_service" } },
            "store": "circuit_state",
            "failureThreshold": 5,
            "failureWindowSeconds": 60,
            "resetTimeoutSeconds": 30
        }"#;

        let spec: CircuitBreakerSpec = serde_json::from_str(json).unwrap();
        assert_eq!(spec.name, "payment_api");
        assert_eq!(spec.failure_threshold, 5);
        assert_eq!(spec.success_threshold, 1); // default
    }

    #[test]
    fn test_parse_saga_spec() {
        let json = r#"{
            "steps": [
                {
                    "id": "step_0",
                    "name": "reserve_inventory",
                    "action": { "tool": { "name": "reserve" } },
                    "compensate": { "tool": { "name": "release" } },
                    "input": { "input": { "path": "$" } }
                }
            ],
            "sagaIdPath": "$.orderId",
            "timeoutMs": 300000
        }"#;

        let spec: SagaSpec = serde_json::from_str(json).unwrap();
        assert_eq!(spec.steps.len(), 1);
        assert_eq!(spec.saga_id_path, Some("$.orderId".to_string()));
    }
}
